torchtext/transforms.py

```diff
@@ -189,3 +193,350 @@ def forward(self, input: Any) -> Any:
         :rtype: Union[List[Union[str, int]], List[List[Union[str, int]]]]
         """
         return F.truncate(input, self.max_seq_len)
+
+
+class GPT2BPETokenizer(Module):
+    """
+    Transform for GPT-2 BPE Tokenizer. 
+
+    Reimplements openai GPT-2 BPE in TorchScript. Original openai implementation 
+    https://github.com/openai/gpt-2/blob/master/src/encoder.py
+
+    :param bpe_encoder: A dict mapping from bpe tokens(e.g sub-words) to
+            bpe token ids. For example: {"ingly": 4420}
+    :type bpe_encoder: Dict[str, int]
+    :param bpe_merge_ranks: A dict mapping a pair of bpe tokens to rank.
+            Torchscript doesn't support tuple as dict key type, concat
+            a pair of string with SEPERATOR (e.g \u0001) as workaround.
+            For example: {"ing\u0001ly": 210}
+    :type bpe_merge_ranks: Dict[str, int]
+    :param byte_encoder: A dict mapping a byte to an unicode character.
+    :type byte_encoder: Dict[int, str]
+    """
+    SEPERATOR: str = "\u0001"
+    LSTRIP_PATTERN: str = "\u0120"
```

Comment:
Class attributes are tricky here. If anyone needs to use custom separator/lstip_pattern, then one needs to modify them on class.

```python
GPT2BPETokenizer.SEPARATOR = my_separator
GPT2BPETokenizer.LSTRIP_PATTERN = my_lstrip_pattern
foo = GPT2BPETokenizer.load_xxx(...)
```

It looks like these separator and strip_pattern are logically connected with the data to be loaded (json file and bpe file), which means that the customization should best happen at loading time.
Making them into class attribute means, that if I have two BPE tokenizers, one with default and the other with custom separator/lstrip_pattern, I cannot use them at the same time.

Looking at the implementation of `get_pairs` and `bpe` method, `self.merge_seperator` are used. This means that they are better coupled with instance, instead of class.

Also this approach is forcing all the `load_XXX` functions to be `classmethod`. In general in Python, there is not much advantage of using `classmethod`/`staticmethod` over plain module level function (except logical grouping of functions and the very specific case where the method is intended to be subclassed and overridden from child class).

Having class method could be troublesome if the instance is going to be used in subprocess (i.e. the instances are pickled and unpicked when passed from main process to subprocess. I am not sure that's the case for this class, but in general, it's better not to have one).

Therefore, if these attributes are often customized, then I recommend to make them as optional constructor argument, else, putting them as a global variable wouldn't hurt much.

Reply:
@mthrok Thanks for the detailed suggestions. Your points make a lot of sense and I will try to get rid of the unnecessary class methods. 

 One thing, I would like to emphasize that these class attributes are not designed to be customizable. The LSTRIP_PATTERN is how GPT-2 encodes spaces. We don't need users to change this unless they want to try out an idea that is different from the original implementation. As far as SEPERATOR is concerned, that is our workaround because Torchscript does not support Tuple as Dict keys. SEPERATOR in that way is closely toed to our implementation and we don't expect any user to modify this. 

> Also this approach is forcing all the load_XXX functions to be classmethod. In general in Python, there is not much advantage of using classmethod/staticmethod over plain module level function (except logical grouping of functions and the very specific case where the method is intended to be subclassed and overridden from child class).

I am not sure I understand why this forces us to make class methods. Let me refactor these to be ScriptModule constants as described at the bottom of https://pytorch.org/docs/stable/jit_language_reference.html